### ISSUE TITLE
Support for Nested Functions

### ISSUE THREAD

**Issue by Shaikh-Ubaid** [lcompilers/lpython#2288]
In the C, C++ and WASM Backends, I think we do not have support for nested functions. Shall we write an ASR->ASR pass that transforms the nested functions structure to a flat structure? In the pass, I think we need to pass the parent function arguments and parent local variables to its nested function with the intent `inout`. We need to update the nested function(s) definition and calls to it accordingly.

**Comment by certik** [comment]
We already have such a nested function pass: https://github.com/lfortran/lfortran/blob/403aa87b819a441e9c6503339dbc8940569bee6b/src/libasr/pass/nested_vars.cpp, so you should just call it in WASM (and other backends as needed). If it doesn't work, then let's fix the bugs.

**Comment by Shaikh-Ubaid** [comment]
The `nested_vars` pass is good. A blocker is that the pass does not flatten the nested structure of functions. It only makes the variables from parent function accessible to the nested functions. Currently in the WASM backend and in the C Backend (here https://github.com/lfortran/lfortran/pull/2233) we explicitly flatten the functions. I think the drawback/limitation at the moment is that in the WASM and C Backends, we might experience function collision/redeclaration when the nested function has the same name as that of a function declared/defined in the global scope.

Shall we have a separate pass that flattens the nested structure of functions to a stacked structure appropriately updating the nested function names so that it does not conflict with any functions defined in the global scope?

**Comment by certik** [comment]
Yes, I think we should. Then the backends don't have to handle any nested functions at all, only single level functions in a module.

**Comment by faze-geek** [comment]
Related https://github.com/lcompilers/lpython/issues/1631
